Grbl_Esp32/src/Pin.cpp

```diff
@@ -0,0 +1,166 @@
+#include "Pins/VoidPinDetail.h"
+#include "Pins/GPIOPinDetail.h"
+#include "Pins/I2SPinDetail.h"
+
+#if defined PIN_DEBUG && defined ESP32
+#    include "Pins/DebugPinDetail.h"
+#endif
+
+#ifdef ESP32
+#    include "Grbl.h"  // grbl_sendf
+#endif
+
+bool Pin::parse(String str, Pins::PinDetail*& pinImplementation) {
+    // Initialize pinImplementation first! Callers might want to delete it, and we don't want a random pointer.
+    pinImplementation = nullptr;
+
+    // Parse the definition: [GPIO].[pinNumber]:[attributes]
+
+    // Skip whitespaces at the start
+    auto nameStart = str.begin();
+    for (; nameStart != str.end() && ::isspace(*nameStart); ++nameStart) {}
+
+    if (nameStart == str.end()) {
+        // Re-use undefined pins happens in 'create':
+        pinImplementation = new Pins::VoidPinDetail();
+        return true;
+    }
+
+    auto idx = nameStart;
+    for (; idx != str.end() && *idx != '.' && *idx != ':'; ++idx) {
+        *idx = char(::tolower(*idx));
+    }
+
+    String prefix = str.substring(0, int(idx - str.begin()));
+
+    if (idx != str.end()) {  // skip '.'
+        ++idx;
+    }
+
+    int pinNumber = 0;
+    if (prefix != "") {
+        if (idx == str.end()) {
+            // Incorrect pin definition.
+            return false;
+        }
+
+        for (; idx != str.end() && *idx >= '0' && *idx <= '9'; ++idx) {
+            pinNumber = pinNumber * 10 + int(*idx - '0');
+        }
+        if (pinNumber < 0 || pinNumber > 253) {
+            // Pin number has to be between [0,253].
+            return false;
+        }
+    }
+
+    String options;
+    if (idx != str.end()) {
+        if (*idx != ':') {
+            // Pin definition attributes or EOF expected.
+            return false;
+        }
+        ++idx;
+
+        options = str.substring(int(idx - str.begin()));
+    }
+
+    // What would be a simple, practical way to parse the options? I figured, why not
+    // just use the C-style string, convert it to lower case, and change the separators
+    // into 'nul' tokens. We then pass the number of 'nul' tokens, and the first char*
+    // which is pretty easy to parse.
+
+    // Build an options parser:
+    Pins::PinOptionsParser parser(options.begin(), options.end());
+
+    // Build this pin:
+    if (prefix == "gpio") {
+        pinImplementation = new Pins::GPIOPinDetail(uint8_t(pinNumber), parser);
+    } else if (prefix == "i2s") {
+#ifdef ESP32
+        pinImplementation = new Pins::I2SPinDetail(uint8_t(pinNumber), parser);
+#else
+        return false;  // not supported
+#endif
+    } else if (prefix == "void") {
+        // Note: having multiple void pins has its uses for debugging. Note that using
+        // when doing 'x == Pin::UNDEFINED' will evaluate to 'false' if the pin number
+        // is not 0.
+        pinImplementation = new Pins::VoidPinDetail(uint8_t(pinNumber));
+    } else {
+#ifdef ESP32
+        grbl_sendf(CLIENT_ALL, "Unknown prefix: " % s "\r\n", prefix.c_str());
```

Comment:
grbl_sendf(CLIENT_ALL, "Unknown prefix: \"%s\"\r\n", prefix.c_str());
intent was to have  quotes around the prefix. as is, won't compile due to space in % s

Reply:
I shouldn't blindly copy-paste things. 

I didn't notice because I test my code on x64 most of the times, and attempt get everything working before moving back to ESP32. Anyways, should be fixed now.